### PR TITLE
QALI-289 Create the zip and zipLast operators to join uncorrelated so…

### PR DESCRIPTION
**Pull request by p-bril**
…urces

### CODE REVIEW

File: factory/src/main/kotlin/io/qalipsis/core/factory/steps/zip/RightSources.kt [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import io.qalipsis.api.context.CorrelationRecord
+import io.qalipsis.api.context.StepName
+import io.qalipsis.api.messaging.Topic
+
+/**
+ * Specifies a remote [io.qalipsis.core.factory.steps.decorators.OutputTopicStepDecorator] provided data.
+ *
+ * @author Polina Bril
+ */
+internal class RightSources<T : Any>(
```
Comment (ericjesse) [inline factory/src/main/kotlin/io/qalipsis/core/factory/steps/zip/RightSources.kt:31]:
```suggestion
internal class RightSource<T : Any>(
```

File: factory/src/main/kotlin/io/qalipsis/core/factory/steps/zip/ZipStep.kt [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        val stepName = this.name
+        rightSources.forEach { source ->
+            consumptionJobs.add(
+                coroutineScope.launch {
+                    log.debug { "Starting the coroutine buffering right records from step ${source.sourceStepName}" }
+                    val subscription = source.topic.subscribe(stepName)
+                    while (subscription.isActive()) {
+                        val record = subscription.pollValue()
+                        log.trace { "Adding right record to channel" }
+                        coroutineScope.launch { channel.send(mapOf(record.stepName to record.value)) }
```
Comment (ericjesse) [inline factory/src/main/kotlin/io/qalipsis/core/factory/steps/zip/ZipStep.kt:75]:
```suggestion
                        rightValues.send(record.value)
```

File: factory/src/main/kotlin/io/qalipsis/core/factory/steps/zip/ZipStep.kt [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        rightSources.forEach { it.topic.close() }
+        channel.cancel()
+    }
+
+    override suspend fun execute(context: StepContext<I, O>) {
+        if (!running) throw NotInitializedStepException()
+
+        val input = context.receive()
+        val latch = Latch(true, "zip-step-${name}")
+        coroutineScope.launch {
```
Comment (ericjesse) [inline factory/src/main/kotlin/io/qalipsis/core/factory/steps/zip/ZipStep.kt:96]:
```suggestion
```
You already are in a suspended scope, no need to start another coroutine in the present case.

File: factory/src/main/kotlin/io/qalipsis/core/factory/steps/zip/ZipStep.kt [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+            val secondaryValues = channel.receive() as Map<StepName, Any?>
+            log.trace { "Forwarding values" }
+            context.send(outputSupplier(input, secondaryValues))
+            latch.release()
+        }
+        latch.await()
+    }
+
+    companion object {
+        @JvmStatic
```
Comment (ericjesse) [inline factory/src/main/kotlin/io/qalipsis/core/factory/steps/zip/ZipStep.kt:106]:
```suggestion
```

File: factory/src/main/kotlin/io/qalipsis/core/factory/steps/zip/ZipStep.kt [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+    override suspend fun execute(context: StepContext<I, O>) {
+        if (!running) throw NotInitializedStepException()
+
+        val input = context.receive()
+        val latch = Latch(true, "zip-step-${name}")
+        coroutineScope.launch {
+            val secondaryValues = channel.receive() as Map<StepName, Any?>
+            log.trace { "Forwarding values" }
+            context.send(outputSupplier(input, secondaryValues))
```
Comment (ericjesse) [inline factory/src/main/kotlin/io/qalipsis/core/factory/steps/zip/ZipStep.kt:99]:
```suggestion
            context.send(input to secondaryValues)
```

File: factory/src/main/kotlin/io/qalipsis/core/factory/steps/zipLast/ZipLastStep.kt [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+ * parents, only one is considered at the primary, from which will the current one will inherit the context.
+ *
+ * This step uses the same mechanism than [io.qalipsis.core.factory.steps.singleton.SingletonProxyStep] to receive
+ * data from secondary steps.
+ *
+ * The step execution is suspended until values from all parent steps are provided.
+ *
+ * @author Polina Bril
+ */
+internal class ZipLastStep<I, O>(
```
Comment (ericjesse) [inline factory/src/main/kotlin/io/qalipsis/core/factory/steps/zipLast/ZipLastStep.kt:45]: Apply the changes proposed for the ZipStep.

File: factory/src/main/kotlin/io/qalipsis/core/factory/steps/zip/ZipStep.kt [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+            latch.release()
```
Comment (ericjesse) [inline factory/src/main/kotlin/io/qalipsis/core/factory/steps/zip/ZipStep.kt:100]:
```suggestion
```

**Review by ericjesse — CHANGES_REQUESTED** [review]

File: factory/src/main/kotlin/io/qalipsis/core/factory/steps/zipLast/ZipLastStep.kt [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+import io.qalipsis.api.logging.LoggerHelper.logger
+import io.qalipsis.api.steps.AbstractStep
+import io.qalipsis.api.sync.Slot
+import io.qalipsis.core.exceptions.NotInitializedStepException
+import io.qalipsis.core.factory.steps.zip.RightSource
+import kotlinx.coroutines.*
+
+
+/**
+ * Step to join flows from several [Step]s into a single one. While this steps technically has several
```
Comment (joval2) [inline factory/src/main/kotlin/io/qalipsis/core/factory/steps/zipLast/ZipLastStep.kt:34]: Why does this class have the same documentation with ZipStep ?

**Review by joval2 — APPROVED** [review]
Please rebase this branch onto master and squash all your commits.

**Review by joval2 — APPROVED** [review]

**Pull request merged**